Makefile

```diff
@@ -657,16 +657,15 @@ clean.go: ; $(info $(H) cleaning...)
 #-----------------------------------------------------------------------------
 # Target: docker
 #-----------------------------------------------------------------------------
-.PHONY: artifacts gcs.push.istioctl-all artifacts installgen
+.PHONY: push gcs.push gcs.push.istioctl-all gcs.push.deb artifacts installgen
 
 # for now docker is limited to Linux compiles - why ?
 include tools/istio-docker.mk
 
-# if first part of URL (i.e., hostname) is gcr.io then upload istioctl and deb
-$(if $(findstring gcr.io,$(firstword $(subst /, ,$(HUB)))),$(eval push: gcs.push.istioctl-all gcs.push.deb),)
-
 push: docker.push installgen
 
+gcs.push: gcs.push.istioctl-all gcs.push.deb
```

Comment:
I like this but I think its changing the behavior. Calling `gcs.push` should also trigger `push`. Before the rule was
```
push: gcs.push.istioctl-all gcs.push.deb
push: docker.push installgen
```

So it would do all of them I think.


We could change it to just do the special gcs stuff, but you would need to update the call sites you changed below to do both

Reply:
Actually it didn't do them all, what was happening is that the rule was getting rewritten. I ran into the original issue because the rewrite that was doing GCS pushing was resulting in the docker push *not* happening.

Reply:
Agree with @sushicw. Since the rule was rewritten, the original dependant targets were not being run. The PR includes changes to call sites that had explicitly set the `HUB| to cause the rewrite to call the new `gcs.push` target directly.

Reply:
How is the rule rewritten? I thoguht this would add new dependencies essentially, but I am not a makefile export.
See reproducer:
```
$ cat Makefile
a:
        echo a
b:
        echo b
c: a
c: b
$ make c
echo a
a
echo b
b
```

The two targets you modified need to also push to docker

Reply:
You're right, I think what I actually ran into before was that the new rule was effectively 

```push: gcs.push.istioctl-all gcs.push.deb docker.push installgen```

Since the GCS stuff was failing for me (since I didn't have it configured and didn't need it) execution failed before it got to the final rules. My initial assumption that the rule was being "rewritten" is wrong, and can easily be proven by adding the conditional to your repro example above. 

I agree with your original point that we should have `gcs.push` depend on `push`.

Reply:
re-read the make documentation for `eval` and tried a small snippet to confirm. You are correct that this was adding prerequisites and not replacing them. Resubmitting to capture the original behavior where `gcs.push` depends on push.
Thanks for the catch!

```make
.PHONY: a b c

T=c
$(if $(findstring c, $(T)),$(eval a: c),)

a: b
```
can run with `make -d` to see that both `b` and `c` are evaluated when `T=c`